### PR TITLE
feat: add support for setting initial solutions with cbc

### LINKED ISSUE #55: Give a solution to the solver
Hi again!

Most, if not all, of the solver allow to give a feasible solution to the solver. I have a case where I have a solution that is optimal, but I don’t know the optimality. The solver find as a bound the objective value, but continue to search as it doesn’t find the solution. For some solver, it can also improve the resolution.

Can we add this feature? Any remark? How to manage the case of the solver that doesn’t support it (I suspect the executable interface doesn’t always allow this)?a

### PR DESCRIPTION
**Pull request by KnorpelSenf**
This is a first attempt at adding support for initial solutions. For now, only coin_cbc is supported. Compatibility with more solvers can be added in subsequent pull requests.

For some reason, the added test still fails. It's probably a dumb oversight, but until somebody spots it, I would be happy to receive feedback on whether the API and implementation of this feature go in the right direction. If you have complaints about the naming (should we rename `set_initial_solution` to `initial_solution` or `initial` or `initialize` or `init` or `set_initial` or `with_initial_solution`?), please let me know and I'll change it.

Towards #55.

### CODE REVIEW

File: src/solvers/mod.rs [hunk 1]
```diff
@@ -199,6 +199,15 @@ pub trait SolverModel {
     fn name() -> &'static str;
 }
 
+/// A solver that can take an initial solution to a problem before solving it
+pub trait WithInitialSolution {
+    /// The type of solution to the problem
+    type Solution: Solution;
+
+    /// Sets the initial solution to the problem
+    fn set_initial_solution(self, solution: &Self::Solution) -> Self;
```
Comment (lovasoa) [inline src/solvers/mod.rs:208]: maybe 

```suggestion
    fn with_initial_solution(self, solution: &Self::Solution) -> Self;
```

?

Comment (KnorpelSenf) [reply src/solvers/mod.rs:208]: Agree 100 %, that was my favourite, too

**Comment by lovasoa** [comment]
The design seems good to me

**Comment by KnorpelSenf** [comment]
Awesome! Now I just have to find out why the problem in the test has no solution the second time I compute it

**Comment by TeXitoi** [comment]
> But what if somebody already has an initial solution without having solved the problem via this library?

It’s an important case I think. In my usecase, I have a solution that I know is feasible and good quality.

**Comment by KnorpelSenf** [comment]
> > But what if somebody already has an initial solution without having solved the problem via this library?
> 
> It’s an important case I think. In my usecase, I have a solution that I know is feasible and good quality.

The same is true for me. The question remains if we need a more ergonomic API to construct a solution. Maybe not, because we're going to have a way to give an initial value to each variable? I'm undecided on this so far

**Comment by lovasoa** [comment]
Yes, I think we need a way to build solutions for this to be helpful. Maybe implement `FromIterator<(Variable, Value)>`

**Comment by KnorpelSenf** [comment]
@lovasoa do you want this done in this PR? If no, what else is needed before this can be merged?

**Comment by lovasoa** [comment]
Yes, I think the feature is not very useful without this

File: src/solvers/coin_cbc.rs [hunk 2]
```diff
@@ -166,6 +167,27 @@ impl SolverModel for CoinCbcProblem {
...
+    fn from_iter<T: IntoIterator<Item = (Variable, f64)>>(iter: T) -> Self {
+        CoinCbcSolution {
+            solution: CbcSolution {
+                raw: todo!(),
+                col_solution: todo!(),
+            },
+            solution_vec: iter.into_iter().map(|(_, value)| value).collect(),
+        }
+    }
+}
```
Comment (KnorpelSenf) [inline src/solvers/coin_cbc.rs:189]: @lovasoa I am a bit lost on how to move forward. Did you have something like this in mind? If yes, how can we construct the underlying solution of the solver? The underlying `coin_cbc` crate does not seem to expose any way of creating solutions from scratch. Yet they are the only data type that can be passed to `set_initial_solution`.

Do I first have to go fix `coin_cbc` before this becomes possible?

Comment (lovasoa) [reply src/solvers/coin_cbc.rs:189]: Indeed, I did not have this in mind, and I think this questions our initial approach...

Maybe `with_initial_solution` should instead be generic and take a trait that is implemented both by Solution AND by (var, value) iterators?

Comment (KnorpelSenf) [reply src/solvers/coin_cbc.rs:189]: I'm not really sure how structuring traits differently solves the underlying problem. There is no way to construct an initial solution for `coin_cbc` except by running the solver, so how can it ever work to set initial solutions at all—unless they were obtained from previous runs?

Comment (KnorpelSenf) [reply src/solvers/coin_cbc.rs:189]: I asked about this in https://github.com/KardinalAI/coin_cbc/issues/37

Comment (KnorpelSenf) [reply src/solvers/coin_cbc.rs:189]: I think I'll manage to implement this in a few days thanks to @TeXitoi

File: src/solvers/mod.rs [hunk 3]
```diff
@@ -199,6 +199,12 @@ pub trait SolverModel {
     fn name() -> &'static str;
 }
 
+/// A solver that can take an initial solution to a problem before solving it
+pub trait WithInitialSolution {
+    /// Sets the initial solution to the problem
+    fn with_initial_solution(self, solution: &Vec<(Variable, f64)>) -> Self;
+}
```
Comment (lovasoa) [inline src/solvers/mod.rs:206]: Maybe this ?

```suggestion
pub trait WithInitialSolution {
    /// Sets the initial solution to the problem
    fn with_initial_solution(
        self,
        solution: impl IntoIterator<Item = (Variable, f64)>
    ) -> Self;
}
```

File: src/solvers/coin_cbc.rs [hunk 4]
```diff
@@ -166,6 +166,16 @@ impl SolverModel for CoinCbcProblem {
     }
 }
 
+impl WithInitialSolution for CoinCbcProblem {
+    fn with_initial_solution(mut self, solution: &Vec<(Variable, f64)>) -> Self {
+        for (var, val) in solution {
```
Comment (lovasoa) [inline src/solvers/coin_cbc.rs:171]: If we don't need it to be a vec, we can loosen the type requirements

**Review by lovasoa — CHANGES_REQUESTED** [review]
Great ! I think we should accept an iterator instead of a vec, and then we are good to merge :+1:

**Comment by KnorpelSenf** [comment]
Nice, will do! 

I can imagine that setting each variable individually is not as good as setting the entire solution at once if one has a full solution available. Should we still have a way to set complete initial solutions?

I honestly don't really see how that would be useful because I don't see a way to construct these solutions, but I just wanted to make sure we're on the same page with this:)

File: src/solvers/mod.rs (hunk 3 above)
Comment (KnorpelSenf) [reply src/solvers/mod.rs:206]: I had the exact same thought when I woke up today

**Comment by lovasoa** [comment]
The current implementation looks good to me. I'll merge it as soon as you make the fix.

**Comment by TeXitoi** [comment]
If we could modify a model, it could be useful to use the solution. An example is lexicographic objectives:
 * you create the problem with the first objective
 * you set a constraint "the previous objective must be at least the quality of the solution found"
 * you set the next objective
 * you set the initial solution to the previous found solution

But, as far as I remember, the current interface doesn’t allow this.

**Comment by KnorpelSenf** [comment]
I agree. I would also like to be able to modify problems and then reuse previous solutions from previous runs. But this is out of scope for the current PR. I might look into this in the future.

**Review by lovasoa — APPROVED** [review]

**Comment by lovasoa** [comment]
# :rocket:

**Pull request merged**